### PR TITLE
V0.2 rl v2 maddpg

### PR DESCRIPTION
**Pull request by buptchan**
# Description
Support multi-agent algorithm `MADDPG` in rl_v2.
Fit experience data recording & sampling for multi-agent scenario.

Test Results:
Algo | Topo | Shortage
:-: | :-: | :-: 
AC | 4p_ssdd_l0.0 | 326665
AC | 4p_ssdd_l0.8 | 958303
MADDPG | 4p_ssdd_l0.0 | 109508
MADDPG | 4p_ssdd_l0.8 | 313367

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [x] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [x] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [x] Linux
- Python version:
  - [ ] 3.6
  - [x] 3.7
- Key information snapshot(s):

## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [x] Add/update the related comments
- [ ] Add/update the related tests
- [ ] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/rl/policy_v2/policy_base.py [hunk 1]
```diff
@@ -395,15 +395,20 @@ def __init__(self, name: str, device: str) -> None:
         """
         super(MultiRLPolicy, self).__init__(name=name, device=device)
 
-    def _shape_check(self, states: np.ndarray, actions: Union[None, List[np.ndarray]]) -> bool:
-        if not (states.shape[0] > 0 and match_shape(states, (None, self.state_dim))):
-            return False
+    def __call__(self, states: List[np.ndarray], agent_ids: List[int]):
+        assert self._shape_check(states=states, actions=None)
+        ret = self._call_impl(states, agent_ids)
```
Comment (lihuoran) [inline maro/rl/policy_v2/policy_base.py:400]: `AbsRLPolicy._call_impl()` does not have parameter `agent_ids`. Personally, I think we need a new abstraction for multi RL policies, for example: `AbsMultiPolicy => AbsMultiRLPolicy => ...`

File: maro/rl/policy_v2/policy_base.py [hunk 2]
```diff
@@ -395,15 +395,20 @@ def __init__(self, name: str, device: str) -> None:
...
+        # assert self._call_post_check(states=states, ret=ret)  # TODO
+        return ret
+
+    def _shape_check(self, states: List[np.ndarray], actions: Union[None, List[torch.Tensor]]) -> bool:
```
Comment (lihuoran) [inline maro/rl/policy_v2/policy_base.py:404]: What is the data type of `actions`? `ndarray` or `Tensor`?

File: maro/rl/policy_v2/policy_base.py [hunk 3]
```diff
@@ -414,22 +419,22 @@ def action_dims(self) -> List[int]:
...
-    def get_actions(self, states: Union[np.ndarray, List[np.ndarray]]) -> List[np.ndarray]:
+    def get_actions(self, states: Union[np.ndarray, List[np.ndarray]], agent_ids: List[int]) -> List[np.ndarray]:
         assert self._shape_check(states=states, actions=None)
-        actions = self._get_actions_impl(states)
+        actions = self._get_actions_impl(states, agent_ids)
         assert self._shape_check(states=states, actions=actions)
         return actions
 
     @abstractmethod
     def _get_actions_impl(self, states: Union[np.ndarray, List[np.ndarray]]) -> List[np.ndarray]:
```
Comment (lihuoran) [inline maro/rl/policy_v2/policy_base.py:439]: Missing parameter `agent_ids` in `_get_actions_impl()`.

File: maro/rl/modeling_v2/critic_model.py [hunk 4]
```diff
@@ -194,15 +194,14 @@ def __init__(self, state_dim: int, action_num: int) -> None:
...
-        return q_matrix.gather(dim=1, index=actions).reshape(-1)
+        q_values = self.q_critic_for_all_actions(states, actions)  # [batch_size, action_num]
+        return q_values.reshape(-1)
 
     @property
     def action_num(self) -> int:
         return self._action_num
 
-    def q_critic_for_all_actions(self, states: torch.Tensor) -> torch.Tensor:
+    def q_critic_for_all_actions(self, states: torch.Tensor, actions: torch.Tensor) -> torch.Tensor:
```
Comment (lihuoran) [inline maro/rl/modeling_v2/critic_model.py:204]: Inconsistency between function name and function content. `q_critic_for_all_actions` means returning Q values for all possible actions according to the given state, so it should not take actions as inputs.

File: examples/rl/cim/policies_v2.py [hunk 5]
```diff
@@ -119,5 +191,9 @@ def _forward_unimplemented(self, *input: Any) -> None:
     policy_func_dict = {
         f"{algorithm}.{i}": lambda name: DiscreteActorCritic(name, MyACNet(), **ac_conf) for i in range(num_policies)
     }
+elif algorithm == "maddpg":
+    agent_nets = [MyActorNet() for i in range(num_agents)]
+    maddpg_policy = MultiDiscreteActorCritic(algorithm, agent_nets, MyCentralizedCriticNet(), **maddpg_conf)
```
Comment (lihuoran) [inline examples/rl/cim/policies_v2.py:196]: Expected type `MultiDiscreteQCriticNetwork`, got `MyCentralizedCriticNet` instead

File: maro/rl/policy_v2/maddpg.py [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+            net.eval()
+
+        with torch.no_grad():
+            actions = []
+            for agent_id, state in zip(agent_ids, input_states):  # iterate `num_sub_agent` times
+                net = self._agent_nets[agent_id]
+                state = torch.from_numpy(state).unsqueeze(0).to(self._device)
+                action, _ = net.get_actions_and_logps(state, self._exploring)  # [batch_size], [batch_size]
+                actions.append(action)
+        return actions
```
[the remark this answers is not in the document]
Comment (buptchan) [reply maro/rl/policy_v2/maddpg.py:184]: List[torch.Tensor]

File: maro/rl/policy_v2/maddpg.py [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+        rewards = torch.from_numpy(rewards).to(self._device)  # [batch_size]
+        terminals = torch.from_numpy(terminals).float().to(self._device)  # [batch_size]
+
+        # critic loss
+        with torch.no_grad():
+            next_actions = [agent.get_actions_and_logps(
+                torch.from_numpy(sub_batch["next_states"]).to(self._device), self._exploring)[0]
+                for agent, sub_batch in zip(self._target_agent_nets, batch)]
+            next_action_tensor = torch.stack(next_actions).T  # [batch_size, sub_agent_num]
+            next_q_values = self._target_critic_net.q_critic(states, next_action_tensor)
```
[the remark this answers is not in the document]
Comment (buptchan) [reply maro/rl/policy_v2/maddpg.py:267]: Bug fixed. Should be `next_states` here.

File: maro/rl/policy_v2/maddpg.py [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+        self.num_epochs = num_epochs
+        self.update_target_every = update_target_every
+        self._min_logp = min_logp
+        self._critic_loss_func = critic_loss_cls() if critic_loss_cls is not None else torch.nn.MSELoss()
+        self._critic_loss_coef = critic_loss_coef
+        self._soft_update_coef = soft_update_coef
+        self._clip_ratio = clip_ratio
+        self._lam = lam
+
+        self._critic_net_version = 0
```
[the remark this answers is not in the document]
Comment (buptchan) [reply maro/rl/policy_v2/maddpg.py:112]: I follow the naming of `DDPG` here. 
If we have to rename the variables, we could do it in another PR.

File: maro/rl/modeling_v2/critic_model.py (hunk 4 above)
Comment (buptchan) [reply maro/rl/modeling_v2/critic_model.py:207]: Added a new interface `q_critic` and restored this one.

**Pull request closed without merging**